vshard/storage/schema.lua

```diff
@@ -186,8 +191,10 @@ local function schema_upgrade(target_version, username, password)
                         next_version)
                     exports = lvexports.compile(exports)
                     handler.upgrade()
-                    lvexports.deploy_funcs(exports)
-                    lvexports.deploy_privs(exports, username)
+                    if M.management_mode == 'auto' then
```

Comment:
The whole `pcall` should be inside if statement. Otherwise, you set `vshard_version` to `0.1.16` without any changes.

But it seems, we should exit earlier, right after `schema_init_0_1_15_0`

Reply:
> Otherwise, you set vshard_version to 0.1.16 without any changes.

Yes. This is intentional (except that there will be a change for some really old vshards - removal of `'oncevshard:storage:1'`). The reason is that when you voluntary choose to manage the access (!= schema) on your own, then you do that outside of the internal version upgrades. Otherwise I would be unable to do any upgrades at all, including potential internal space changes for `_bucket` or creation of some new spaces, when the "users" just want to manage the exports on their own.

> But it seems, we should exit earlier, right after schema_init_0_1_15_0

No, there is no schema management mode called `manual`. There is only `manual_access`. The "users" only have control over the exports. The backbone of the upgrade procedure is owned by vshard. If fully manual schema is needed somewhere, then that place is behaving really wrong. This is like saying that you want to create all system spaces and WAL before Tarantool is started first time. To those users the current exports won't help, and they shouldn't help.

I was thinking though if that would be possible at all to create generic schema version log which could contain everything we have now and what we would potentially need. But it appears it can't be done as easily as the exports. Such a schema log would need to be able to describe untrivial things:
- Deletion and insertion of random keys in random spaces (to drop `'oncevshard:storage:1'` from `_schema`, to update `vshard_version`);
- Creation of spaces and indexes (to create `_bucket` and its indexes);
- Alter of those spaces and indexes (I want to change `_bucket` status field from string to unsigned), with changing their data first to prepare for the new space format (don't even know how to describe it in the form of Lua tables for space whose content is unknown at the time of upgade);

The last point means that we can't even do the schema log in the form of a list of xlog-like rows, because for `_bucket` alter we don't know the number of rows nor their exact keys (stuff changes during rebalancing). It would mean export either in the form of Lua-code wrapped into string, or SQL-code (wouldn't work on 1.10 which is still supported). That solution would make this schema log unusable for anything except just applying it right away (otherwise go parse Lua and SQL to extract details), which leads to the point why not just let vshard do the same and skip this schema log.

So my resolution was that I simply don't want to export that, because it would be too hard to use, and I see no reason why it should be used and maintained. In my tests in vtest somehow even before any function-export logs I was able to start and configure clusters which is supposedly what `config` also wants to do (I only do understand the manual user management, not exports management).

Nonetheless the patch exists because exporting just the functions and users and privs is easy enough not to bother with arguing about them.